### PR TITLE
Make allocation/free/clean available to providers

### PR DESCRIPTION
**Pull request by richsalz**
As a side-note, editing crypto/build.info for the FIPS provider, as opposed to providers/fips/build.info, is kind of confusing.

This DOES NOT WORK if an application calls ```CRYPTO_set_mem_functions```  I will do so in a separate PR once this is merged.

### CODE REVIEW

File: include/openssl/core_numbers.h [hunk 1]
```diff
@@ -58,6 +63,53 @@ OSSL_CORE_MAKE_FUNC(const OSSL_ITEM *,
...
+#define OSSL_FUNC_CORE_GET_CRYPTO_SECURE_ZALLOC 20
+OSSL_CORE_MAKE_FUNC(void *,
+        CRYPTO_secure_zalloc, (size_t num, const char *file, int line))
+#define OSSL_FUNC_CORE_GET_CRYPTO_SECURE_FREE   21
+OSSL_CORE_MAKE_FUNC(void,
+        CRYPTO_secure_free, (void *ptr, const char *file, int line))
+#define OSSL_FUNC_CORE_GET_CRYPTO_SECURE_CLEAR_FREE 22
+OSSL_CORE_MAKE_FUNC(void,
+        CRYPTO_secure_clear_free, (void *ptr, size_t num, const char *file, int line))
+#define OSSL_FUNC_CORE_GET_CRYPTO_SECURE_MALLOC_INITIALIZED 23
```
Comment (paulidale) [inline include/openssl/core_numbers.h]: Is this one necessary?  It's not the providers concern if an application has started secure memory or not.

Comment (richsalz) [reply include/openssl/core_numbers.h]: I waffled on this, but I ultimately decided that a provider does have the right to know how it's keys are stored. But I don't have strong feelings either way.

Comment (paulidale) [reply include/openssl/core_numbers.h]: Thoughts @openssl/committers ?

I'm undecided too.  I'm leaning towards omitting it and adding it when and if it becomes desirable.  Primarily because a public interface can't be removed but a private one can be made public.

Comment (mattcaswell) [reply include/openssl/core_numbers.h]: I lean towards @paulidale's view.

Comment (richsalz) [reply include/openssl/core_numbers.h]: Okay, I removed it and pushed a new commit.

Comment (paulidale) [reply include/openssl/core_numbers.h]: Thanks.

**Comment by paulidale** [comment]
Ping @openssl/committers ?

File: crypto/provider_core.c [hunk 2]
```diff
@@ -643,6 +643,21 @@ static int core_get_params(const OSSL_PROVIDER *prov, const OSSL_PARAM params[])
 static const OSSL_DISPATCH core_dispatch_[] = {
     { OSSL_FUNC_CORE_GET_PARAM_TYPES, (void (*)(void))core_get_param_types },
     { OSSL_FUNC_CORE_GET_PARAMS, (void (*)(void))core_get_params },
+
```
Comment (slontis) [inline crypto/provider_core.c]: NIT: maybe remove blank line

File: providers/fips/fipsprov.c [hunk 3]
```diff
@@ -89,12 +103,124 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
...
+{
+    c_CRYPTO_clear_free(ptr, num, file, line);
+}
+
+void *CRYPTO_realloc(void *addr, size_t num, const char *file, int line)
+{
+    return c_CRYPTO_realloc(addr, num, file, line);
+}
+
+void *CRYPTO_clear_realloc(void *addr, size_t old_num, size_t num, const char *file, int line)
```
Comment (slontis) [inline providers/fips/fipsprov.c]: NIT: Line looks too long

File: crypto/provider_core.c (hunk 2 above)
Comment (richsalz) [reply crypto/provider_core.c]: kept the blank line, added a comment why.

File: providers/fips/fipsprov.c (hunk 3 above)
Comment (richsalz) [reply providers/fips/fipsprov.c]: fixed.

**Comment by paulidale** [comment]
A second review is required.

**Comment by richsalz** [comment]
your colleague @slontis approved. :)  maybe he wants to do this as his first official commit?

**Comment by t-j-h** [comment]
We are updating the tooling to make sure same-organisation approvals aren't seen as a match to avoid any potential **perceived** conflict of interest in the review process ...

**Comment by richsalz** [comment]
@t-j-h nice to see that being made policy (since i first started it about akamai).  Just for clarification, does that also mean not approving PR's from one's one organization?

**Comment by slontis** [comment]
In future I will not approve if @paulidale has already approved. (This applies to our own PR's also).

**Comment by richsalz** [comment]
Rebased; needs reviews.

**Comment by paulidale** [comment]
Needs a rebase and a second review.

**Comment by paulidale** [comment]
Has a decision about the control functions being offer from providers or not?

**Comment by levitte** [comment]
I will continue to point out that `set_params` and `get_params` type functions can be used as controls. It's all in what the receiving end does in response to OSSL_PARAM values...

**Comment by slontis** [comment]
Ping @openssl/committers ?

File: providers/fips/fipsprov.c [hunk 4]
```diff
@@ -22,10 +22,25 @@
...
 
 /* Functions provided by the core */
-static OSSL_core_get_param_types_fn *c_get_param_types = NULL;
-static OSSL_core_get_params_fn *c_get_params = NULL;
-static OSSL_core_put_error_fn *c_put_error = NULL;
-static OSSL_core_add_error_vdata_fn *c_add_error_vdata = NULL;
+static OSSL_core_get_param_types_fn *c_get_param_types;
+static OSSL_core_get_params_fn *c_get_params;
+static OSSL_core_put_error_fn *c_put_error;
+static OSSL_core_add_error_vdata_fn *c_add_error_vdata;
```
Comment (levitte) [inline providers/fips/fipsprov.c]: Why did you remove the initialization?

Comment (slontis) [reply providers/fips/fipsprov.c]: That was a seperate PR .

**Comment by levitte** [comment]
Regarding dispatch macro names, @richsalz and I have already had a discussion in #8887 and disagree somewhat.  At the very least, I think that `_GET_` in the middle of them unnecessary.

Also, for naming purposes, I'm thinking that if we do want to have something that is explicitly meant to provide the public OpenSSL API, the dispatch macro names for them should maybe have a different prefix than `OSSL_FUNC_CORE`...  perhaps `OSSL_FUNC_OSSL`?  So, something like this?

``` C
#define OSSL_FUNC_OSSL_CRYPTO_MALLOC       nn
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_malloc, (size_t num, const char *file, int line))
```

**Comment by slontis** [comment]
Maybe @levitte & @mattcaswell should confer on these names so it can move forward.

**Comment by levitte** [comment]
We had a quick chat, and @mattcaswell said something that reminded me to look more closely in core_numbers.h and remind myself.
There's this comment:

``` C
So revisiting this, there's this comment in core_numbers.h:

``` C
/*-
 * Identities
 * ----------
 *
 * All series start with 1, to allow 0 to be an array terminator.
 * For any FUNC identity, we also provide a function signature typedef
 * and a static inline function to extract a function pointer from a
 * OSSL_DISPATCH element in a type safe manner.
 *
 * Names:
 * for any function base name 'foo' (uppercase form 'FOO'), we will have
 * the following:
 * - a macro for the identity with the name OSSL_FUNC_'FOO' or derivates
 *   thereof (to be specified further down)
 * - a function signature typedef with the name OSSL_'foo'_fn
 * - a function pointer extractor function with the name OSSL_'foo'
 */
```

So logically, if you want the core to offer `ERR_put_error` and you want to macro name, signature and all that to match, it should be like this:

``` C
/*-
 * Identities
 * ----------
 *
 * All series start with 1, to allow 0 to be an array terminator.
 * For any FUNC identity, we also provide a function signature typedef
 * and a static inline function to extract a function pointer from a
 * OSSL_DISPATCH element in a type safe manner.
 *
 * Names:
 * for any function base name 'foo' (uppercase form 'FOO'), we will have
 * the following:
 * - a macro for the identity with the name OSSL_FUNC_'FOO' or derivates
 *   thereof (to be specified further down)
 * - a function signature typedef with the name OSSL_'foo'_fn
 * - a function pointer extractor function with the name OSSL_'foo'
 */
```

Following that, and if the intention is to have a 1:1 mapping to the public API, then it should look like this, right?

``` C
/* Memory allocation, freeing, clearing. */
#define OSSL_FUNC_CRYPTO_MALLOC       10
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_malloc, (size_t num, const char *file, int line))
#define OSSL_FUNC_CRYPTO_ZALLOC       11
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_zalloc, (size_t num, const char *file, int line))
#define OSSL_FUNC_CRYPTO_MEMDUP       12
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_memdup, (const void *str, size_t siz, const char *file, int line))
#define OSSL_FUNC_CRYPTO_STRDUP       13
OSSL_CORE_MAKE_FUNC(char *,
        CRYPTO_strdup, (const char *str, const char *file, int line))
#define OSSL_FUNC_CRYPTO_STRNDUP      14
OSSL_CORE_MAKE_FUNC(char *,
        CRYPTO_strndup, (const char *str, size_t s, const char *file, int line))
#define OSSL_FUNC_CRYPTO_FREE         15
OSSL_CORE_MAKE_FUNC(void,
        CRYPTO_free, (void *ptr, const char *file, int line))
#define OSSL_FUNC_CRYPTO_CLEAR_FREE   16
OSSL_CORE_MAKE_FUNC(void,
        CRYPTO_clear_free, (void *ptr, size_t num, const char *file, int line))
#define OSSL_FUNC_CRYPTO_REALLOC      17
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_realloc, (void *addr, size_t num, const char *file, int line))
#define OSSL_FUNC_CRYPTO_CLEAR_REALLOC 18
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_clear_realloc, (void *addr, size_t old_num, size_t num, const char *file, int line))
#define OSSL_FUNC_CRYPTO_SECURE_MALLOC 19
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_secure_malloc, (size_t num, const char *file, int line))
#define OSSL_FUNC_CRYPTO_SECURE_ZALLOC 20
OSSL_CORE_MAKE_FUNC(void *,
        CRYPTO_secure_zalloc, (size_t num, const char *file, int line))
#define OSSL_FUNC_CRYPTO_SECURE_FREE   21
OSSL_CORE_MAKE_FUNC(void,
        CRYPTO_secure_free, (void *ptr, const char *file, int line))
#define OSSL_FUNC_CRYPTO_SECURE_CLEAR_FREE 22
OSSL_CORE_MAKE_FUNC(void,
        CRYPTO_secure_clear_free, (void *ptr, size_t num, const char *file, int line))
#define OSSL_FUNC_OPENSSL_CLEANSE 23
OSSL_CORE_MAKE_FUNC(void,
        OPENSSL_cleanse, (void *ptr, size_t len))
# define OSSL_FUNC_OPENSSL_HEXSTR2BUF 24
OSSL_CORE_MAKE_FUNC(unsigned char *,
        OPENSSL_hexstr2buf, (const char *str, long *len))
```

**Comment by levitte** [comment]
(previous comment edited, I had pasted in much more than I intended...)

File: providers/fips/fipsprov.c (hunk 4 above)
Comment (richsalz) [reply providers/fips/fipsprov.c]: Because of the C guarantee to initialize to NULL.

**Comment by richsalz** [comment]
@levitte regarding https://github.com/openssl/openssl/pull/8886#issuecomment-496207559 isn't that what this PR has? But on the other hand, using one person's code comment to justify a policy seems a little weak to me :)

**Comment by richsalz** [comment]
Re https://github.com/openssl/openssl/pull/8886#issuecomment-496180499, perhaps the OMC, the committers, and/or the FIPS sponsors should be included in any such discussions as how to name things?  Not just two paid fellows.

**Comment by richsalz** [comment]
I fixed the define names thanks for the fix!

Can I suggest that maybe in the future we don't need to line up the values with things like this, lots of other lines between defines?  :)  Just put name space value

**Comment by richsalz** [comment]
The function and define names were fixed, this is rebased, so now it needs a reconfirm and another review.

**Comment by paulidale** [comment]
CI failures look relevant.

**Comment by richsalz** [comment]
Oops, clobbered some of my fixes.  New commit pushed that fixes the CI (and function/macro naming) issues.  There is now a new open issue, however, that is described in https://github.com/openssl/openssl/commit/db3480eb18c9889f5ec28000e8fcc9a94eee6554

**Comment by richsalz** [comment]
argh, rebased (that was a hard one) and pushed.  ping.

**Comment by richsalz** [comment]
I cannot figure out why some of the CI builds are failing; time-out issues?  Any clues?

**Comment by paulidale** [comment]
The one I looked at is the evp_extra test failing -- I don't think this one isn't timing critical.
Hmmm...

**Comment by slontis** [comment]
I also was looking at that same test..
This is probably not helpful...
See if you can reproduce it locally with your changes.. (Maybe it only happens with non debug?)

>export LD_LIBRARY_PATH=.
>export OPENSSL_MODULES=providers
>./test/evp_extra_test

If you know which test it is you can do this
>./test/evp_extra_test -test 13 -iter 4

**Comment by slontis** [comment]
Hmm maybe one of your function signatures is wrong.. If you call a function with incorrect args it can just crash.

File: providers/fips/fipsprov.c [hunk 5]
```diff
@@ -351,3 +408,74 @@ const OSSL_PROVIDER *FIPS_get_provider(OPENSSL_CTX *ctx)
 
     return fgbl->prov;
 }
+
+void *CRYPTO_malloc(size_t num, const char *file, int line)
+{
+    return c_CRYPTO_malloc(num, file, line);
```
Comment (slontis) [inline providers/fips/fipsprov.c]: What happens if these c_CRYPTO_ calls dont get resolved and they are NULL?

Comment (richsalz) [reply providers/fips/fipsprov.c]: yeah, i'm tracing down why ```c_CRYPTO_zalloc``` is NULL.  to answer your question, you crash:
```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt 3
#0  0x0000000000000000 in ?? ()
#1  0x00007ffff6bef492 in CRYPTO_zalloc (num=248, 
    file=0x7ffff6bf3789 "crypto/context.c", line=118)
    at providers/fips/fipsprov.c:419
#2  0x00007ffff6bb80b0 in OPENSSL_CTX_new () at crypto/context.c:118
(More stack frames follow...)
(gdb) p c_CRYPTO_zalloc 
$1 = (OSSL_CRYPTO_zalloc_fn *) 0x0
```

**Comment by richsalz** [comment]
AARGH.  I figured it out.  Line 234:
```
	    OPENSSL_CTX *ctx = OPENSSL_CTX_new();
```
which calls CRYPTO_zalloc, but the upcall hasn't been created yet!

Pinging @mattcaswell and anyone else who has thoughts about how to fix this issue.

**Comment by slontis** [comment]
This is a recent change that moved the CTX_new() to before the resolving takes place..
At least some of the resolve will need to move up before the new(). If there are any failures then maybe the statics should all be set back to NULL.

**Comment by richsalz** [comment]
I can fix the segv, but then there's memory leak that doesn't happen on master. :(  see https://github.com/openssl/openssl/pull/8886/commits/5a3e30c212405d4fde5e9e286c0cf73eb4291126

**Comment by richsalz** [comment]
Doing all the resolves first is reasonable, they are just looking up and setting static globals.

File: providers/fips/fipsprov.c [hunk 6]
```diff
@@ -240,22 +244,71 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
...
         return 0;
+    if ((fgbl = openssl_ctx_get_data(ctx, OPENSSL_CTX_FIPS_PROV_INDEX,
+                                     &fips_prov_ossl_ctx_method)) == NULL) {
+        OPENSSL_CTX_free(ctx);
+        return 0;
+    }
+    fgbl->prov = provider;
 
+    if ((ctx = OPENSSL_CTX_new()) == NULL)
+        return 0;
```
Comment (mattcaswell) [inline providers/fips/fipsprov.c]: This is a bug. We shouldn't be calling OPENSSL_CTX_new() twice (not your fault, its mine, introduced in some bad merge somewhere).

Comment (richsalz) [reply providers/fips/fipsprov.c]: Thanks.  Curious I didn't see the leak on master.  Shrug.  Fixup commit pushed, hoping for the CI's to be green now :)

**Comment by richsalz** [comment]
This is nearly two months old and got first approval more than a month ago. Since then I've rebased multiple times, worked around errors in master, etc.  Right now there's an appveyor build failure which I don't understand: https://ci.appveyor.com/project/openssl/openssl/builds/25709908/job/r85f0x7jbnf6824v#L5017

**Comment by richsalz** [comment]
This builds.  It's needed.  It's been sitting for weeks/months.  Ping.

File: crypto/provider_core.c [hunk 7]
```diff
@@ -805,6 +809,26 @@ static const OSSL_DISPATCH core_dispatch_[] = {
     { OSSL_FUNC_CORE_PUT_ERROR, (void (*)(void))core_put_error },
     { OSSL_FUNC_CORE_ADD_ERROR_VDATA, (void (*)(void))core_add_error_vdata },
 #endif
+
+    { OSSL_FUNC_CRYPTO_MALLOC, (void (*)(void))CRYPTO_malloc },
```
Comment (mspncp) [inline crypto/provider_core.c]: General comment: why isn't there a macro for declaring these entries to reduce boilerplate code? For example, `OSSL_DISPATCH_ENTRY(CRYPTO_MALLOC, CRYPTO_malloc)`?

File: include/openssl/core_numbers.h [hunk 8]
```diff
@@ -35,7 +35,12 @@ extern "C" {
  * - a function pointer extractor function with the name OSSL_'foo'
  */
 
-/* Helper macro to create the function signature typedef and the extractor */
+/*
+ * Helper macro to create the function signature typedef and the extractor
+ * |type| is the return-type of the function, |name| is the name of the
+ * function to fetch, and |args| is a parenthesized list of parameters
+ * for the function (that is, it is |name|'s function signature).
+ */
```
Comment (mspncp) [inline include/openssl/core_numbers.h]: The macro name OSSL_CORE_MAKE_FUNC is not very expressive IMHO.  (It _makes_ a _function_?)

Maybe replace _MAKE_ by _DECLARE_ or _DEFINE_ and find a more specifice wording for _FUNC_?

Something like  OSSL_DECLARE_DISPATCH_FUNCTION_GETTER(type,name,args)?  (just wildly guessing)

**Comment by paulidale** [comment]
I'll leave the last two comments for @mattcaswell (& perhaps @levitte) to respond to.
In the meantime, I've merged this.  Thanks for your patience.

File: include/openssl/core_numbers.h (hunk 8 above)
Comment (paulidale) [reply include/openssl/core_numbers.h]: @mattcaswell @levitte

File: crypto/provider_core.c (hunk 7 above)
Comment (paulidale) [reply crypto/provider_core.c]: @mattcaswell @levitte

File: include/openssl/core_numbers.h (hunk 8 above)
Comment (levitte) [reply include/openssl/core_numbers.h]: I 38 character macro name...  please tell me you're joking, that's half the real estate of one line (80 columns).  Also, "GETTER" is inappropriate, as that's just half of what the macro does...  it also declares a function type.

File: crypto/provider_core.c (hunk 7 above)
Comment (levitte) [reply crypto/provider_core.c]: I don't mind such a macro.  PR welcome :wink:

File: include/openssl/core_numbers.h (hunk 8 above)
Comment (levitte) [reply include/openssl/core_numbers.h]: Regarding "FUNC", we already have that on other macros, so I think it's fitting to use the same in this macro.

Comment (levitte) [reply include/openssl/core_numbers.h]: All in all, I would suggest `OSSL_CORE_MAKE_DISPATCH_FUNC` if anything.  It's still pretty long, but I'll live, and I would guess that the added `DISPATCH` in the name makes it expressive enough.  What do you think, @mspncp?

Comment (levitte) [reply include/openssl/core_numbers.h]: (this isn't really the right place to discuss much further...  I see enough here now to inspire someone to make the appropriate PR, together with an added `OSSL_DISPATCH_ENTRY` as suggested above)

Comment (mspncp) [reply include/openssl/core_numbers.h]: > I 37 character macro name... please tell me you're joking, that's half the real estate of one line (80 columns). Also, "GETTER" is inappropriate, as that's just half of what the macro does... it also declares a function type.

As I wrote, it was only a 'wild guess', a quick example shot from the hip. I know too little about the context to propose the optimal fitting name. But any of the two proposals you made is already much better than "make func".
